bcipy/simulator/sim_copy_phrase.py

```diff
@@ -0,0 +1,148 @@
+
+        self.logger.info(f"Loaded model from {model_file}, with model_class {model_class}")
+
+        target_eeg_evidences, non_target_eeg_evidences = [], []
+        while not self.data_loader.is_end():
+            inquiry_trials, this_inquiry_letters, this_inquiry_labels = self.data_loader.next()
+            target_eeg_evidence, non_target_eeg_evidence = [], []
+            response = model.predict(inquiry_trials, this_inquiry_letters, symbol_set=self.symbol_set)
+
+            symbol_letter_idx_map = [self.symbol_set.index(letter) for letter in this_inquiry_letters]
+            for i, label in enumerate(this_inquiry_labels):
+                if label:
+                    target_eeg_evidence.append(response[symbol_letter_idx_map[i]])
+                else:
+                    non_target_eeg_evidence.append(response[symbol_letter_idx_map[i]])
+
+            target_eeg_evidences.append(target_eeg_evidence)
+            non_target_eeg_evidences.append(non_target_eeg_evidence)
+
+        target_eeg_evidences, non_target_eeg_evidences = np.concatenate(target_eeg_evidences), np.concatenate(
+            non_target_eeg_evidences)
+
+        # Get target and non_target eeg values from actual experiment in session.json for comparison
+        session_records = evidence_records(read_session(self.replay_data_path / SESSION_DATA_FILENAME))
+        session_target_eeg = np.array([record.eeg for record in session_records if record.is_target])
+        session_non_target_eeg = np.array([record.eeg for record in session_records if not record.is_target])
+
+        # note this flattens any grouping of inquiries which might be problem for future visualizations
+
+        # visualizing comparison of results
+        plot_replay_comparison(target_eeg_evidences, non_target_eeg_evidences, session_target_eeg,
+                               session_non_target_eeg, self.save_dir, self.parameters)
+
+        self.logger.info(f"Plotted comparison charts "
+                         f"\n New: {len(target_eeg_evidences)} target_eeg_evidences | {len(non_target_eeg_evidences)} non_target_eeg_evidences "
+                         f"\n Old: {len(session_target_eeg)} target_eeg_evidences | {len(session_non_target_eeg)} non_target_eeg_evidences ")
+
+    def get_param(self, name):
+        return self.parameters.get(name, None)
+
+    def init_data_loader(self):
+
+        if self.replay_data_path:
+            self.data_loader = ReplayDataLoader(self.replay_data_path, self.parameters)
+        else:
+            pass  # TODO sampling data logic
+        pass
+
+    def load_parameters(self, path):
+        # TODO validate parameters
+        parameters = load.load_json_parameters(path, value_cast=True)
+        sim_parameters = load.load_json_parameters(
+            "bcipy/simulator/sim_parameters.json", value_cast=True)
+
+        parameters.add_missing_items(sim_parameters)
+        return parameters
+
+
+class ReplayDataLoader(SimulatorData):
```

Comment:
The naming/hierarchy here is a little odd for me. Essentially it reads that a DataLoader is a Data source. Maybe just remove the 'Loader' part of the name and call it `ReplayData`?

So in this case the data knows how to load itself. But we may want to separate the two concepts of a loader and an iterator. We could have a `DataLoader` abstract class with a `load` method that produces an `Iterator`. Python already has the Iterator protocol so SimulatorData is redundant.

Reply:
Though it occurs to me that while iteration works well for replaying a session, it may not be the best approach for sampling. In that case the data is not sequential, but is more of a lookup that depends on context provided in the arguments. It might be a good idea to try to work out what the API for what Sample based data would like and see if you can come up with an abstract representation that works for both.